### ISSUE TITLE
OS specific dev containers

### ISSUE THREAD

**Issue by Chuxel** [microsoft/vscode-remote-release#2067]
One of the value propositions of dev containers is that teams that use mixed operating systems can all benefit from a consistent toolchain regardless of host OS. However, in some cases, there may be development containers that will only ever work on a specific operating system. (See https://github.com/microsoft/vscode-dev-containers/pull/169.)

This feature request tracks the idea of being dev containers being able to declare where they are expected to work.  

For example, one possible way this could work is that `devcontainer.json` could have a property that declared the host OS platforms it supported (`macos`, `windows`, `linux`).  Not specifying any platform could mean that it works with any for backwards compatibility.  

To use a property like this we could set things up so that, if a devcontainer.json in the vscode-dev-containers repo has a value for property, any lists in VS Code include / exclude it based on the host OS VS Code is running on.  Similarly, in the case of an "open folder in container..." scenario with an existing repository with a devcontainer.json in it, we could at least warn that the dev container is not intended for use on the current host OS if it is not in the list.

**Comment by bhack** [comment]
I cannot see the internals but I suppose that It Is not a so complex task.

**Comment by frank-dspeed** [comment]
This should not get done!!! The Container and The Software your Coding Identifys it self already.

**Comment by frank-dspeed** [comment]
Don't get me wrong but i wan't to give you some examples a -windows image will only run on windows a -arm image will run on arm 

you can not even guess on what platforms this can get executed  i am runnong for example vscode on my android with custom theme. i think no one did expect that?!?

**Comment by frank-dspeed** [comment]
containers have only one boundarie the CPU Architecture they are running on https://github.com/unicorn-engine/unicorn

**Comment by bhack** [comment]
@frank-dspeed We need this cause docker has some feature portability limit and you cannot limit the use cases to the common features. 
Check again the thread https://github.com/microsoft/vscode-dev-containers/pull/169

**Comment by bhack** [comment]
How much work Is required for this feature?

**Comment by frank-dspeed** [comment]
i am for adding that only to the description of the dev container maybe as anything else would mean we need to guess the host os

**Comment by bhack** [comment]
We have `os.platform()` and `os.type()` what is the problem?

**Comment by bhack** [comment]
@Chuxel we need to think also to cross-compile sceneraios. 
E.g. see openwrt Docker images when you have different containers with prepared precompiled toolchains for every arch:
https://hub.docker.com/r/openwrtorg/sdk/tags

**Comment by samrocketman** [comment]
# Use case

Take, for example, the open source game [Endless Sky][es].  It builds across 3 different operating systems.

You can see the [build artifacts workflow][actions] for Windows, Mac, and Linux.

# Multi-OS support

It makes sense to build a windows artifact when developing from a Windows host.  Same goes for Mac on Mac or Linux on Linux.

I know there's no Mac containers but at a minimum it should be possible to support Windows containers on Windows and Linux containers on Linux.

[es]: https://endless-sky.github.io/
[actions]: https://github.com/endless-sky/endless-sky/blob/master/.github/workflows/cd.yaml

**Comment by frank-dspeed** [comment]
Now after some more years of edge case evaluation and triage the result is still that you should simply document the processor architecture that runs on and the os. Containers could also get deployed to remote container hosts while they are containers. So the Host OS of the developer is less relevant. see also vscode online

**Comment by chrmarti** [comment]
Windows containers are tracked as https://github.com/microsoft/vscode-remote-release/issues/445.

FreeBSD support is tracked as https://github.com/microsoft/vscode-remote-release/issues/727.

This issue is about having host OS-specific properties in the devcontainer.json.

**Comment by blackliner** [comment]
What about being able to run two different environments: Ubuntu 18.04 and Ubuntu 20.04

Would this also be covered? Our product gets compiled for different (Linux) OS, and being confined to only one environment in the devcontainer world is limiting :-(

**Comment by frank-dspeed** [comment]
@blackliner out of my view this will not even really be covered by this solution you could define maybe a remote location that is able to offer both environments but thats up to you.

**Comment by chrmarti** [comment]
@blackliner That sounds like you would want different versions of Ubuntu inside the container. (This issue is about the host OS, not the container OS.)

**Comment by blackliner** [comment]
Yes, it's about having different OS inside the docker.

**Comment by Chuxel** [comment]
@blackliner Yeah its quite different since what you are talking about would be a way to affect the image or Dockerfile when starting up a dev container... or put another way having multiple dev container configurations.  We've started on a capability that would allow multiple dev contaienr configurations to be under the same `.devcontainer` folder (https://github.com/microsoft/dev-container-spec/issues/6). 

At that point you can have a common Dockerfile or Docker Compose file with different arguments.  There is then also the [dev container features](https://code.visualstudio.com/docs/remote/containers#_dev-container-features-preview) capability that allows you to reuse scripts and config across multiple dev container configuration files or even repositories. 

What this is about is not any of that, but rather the ability to indicate that certain devcontainer.json properties should only be applied for a given local OS (not the one in the conntainer).

**Comment by blackliner** [comment]
Don't get me wrong, I am fully aware about your proposal being about different Host OS and mapping them to guest OS. But IMHO this is just one view of a fully configurable N to N mapping.
1 host OS can run 1 guest: easy
N host OS can run 1 guest: possible today (kinda, given docker supports your platform)
1 host OS can run N guest: my request
N host OS can run N guest: your request

Make sense? Oh, and thanks for the links, will dig them later.

**Comment by frank-dspeed** [comment]
@blackliner there is something that is a logic failure in your complexity matrix you need to add the CPU to that 1 CPU type can run only (N) Host OS the () references the N as partial as not all Processors can run any OS. 

you need to add first a CPU level to that also the CPU Tells you if docker support is possible not only the OS.  Containers are Layered over the Host OS Kernel running with it and its sys calls to the CPU.

**Comment by frank-dspeed** [comment]
@manuth i get you got not a single usecase for that anymore as that file does not matter as much as it did before

- [ ] https://github.com/microsoft/vscode/issues/167505

when i finished that and it lands you will not even run vscode anymore on your host where your docker is installed you can then simply use the marketplace extension of your host provider to spin up containers 

or use the openstack docker marketplace addons to control remote hosts from your browser.

**Comment by frank-dspeed** [comment]
oh i guess i missed the point this deprecates containers as a whole concept it deprecates libcontainer and kubernetes :D have a nice day.

**Comment by stepkun** [comment]
Currently there are several incompatibilities between docker compose on Window and on Linux if you want to use video and audio from within a container.
Therefore I had to separate the definitions into platform specific compose-<platform>.yaml files.
It would be really great if I could select the proper file winthin devcontainer.json with the help of a built in variable , which provides different values for different platforms, for example like nodes process.platform() does.

So that I could do someting like
´´´
	"dockerComposeFile": [
		"compose-common.yaml",
		"compose-${host_platform}.yaml"
	],
´´´

**Comment by rvodden** [comment]
Another use case here is if you're using rootless docker on Linux with vscode then remote user has to be "root", whereas on windows its preferable to run as a non-root user.

**Comment by stepkun** [comment]
Recently I found a workaraound:
naming the linux specific file `compose-.yaml` (pay attention to the dash before the dot!)
and naming the windows specific file `compose-Windows_NT.yaml`
then in the `devcontainer.json` I use
```
	"dockerComposeFile": [
		// on Linux ${localEnv:OS} will evolve to "" on Window to "Windows_NT"
		"compose-${localEnv:OS}.yaml"
	],
```
